File: src/stimuli/generate_session_sequence_wavs.py
"""Concatenate silence and periodic click trains into long session WAVs."""

from __future__ import annotations

import argparse
import csv
from pathlib import Path
from typing import List, Sequence

import numpy as np

try:
    from .generate_click_stimuli import (
        load_config,
        make_click_kernel,
        periodic_click_train,
        write_wav_mono,
    )
except ImportError:
    from generate_click_stimuli import (
        load_config,
        make_click_kernel,
        periodic_click_train,
        write_wav_mono,
    )


def silence_segment(duration_s: float, sample_rate: int) -> np.ndarray:
    n = int(round(duration_s * sample_rate))
    return np.zeros(n, dtype=np.float32)


def sound_segment(
    freq_hz: float, duration_s: float, sample_rate: int, kernel: np.ndarray
) -> np.ndarray:
    n = int(round(duration_s * sample_rate))
    return periodic_click_train(n, sample_rate, freq_hz, kernel)


def concat_session(
    codes: Sequence[int],
    silence_s: float,
    sound_s: float,
    sample_rate: int,
    click_ms: float,
    amplitude: float,
    rng_seed: int,
) -> np.ndarray:
    np.random.seed(rng_seed)
    kernel = make_click_kernel(click_ms, sample_rate, amplitude)

    chunks: List[np.ndarray] = []
    for code in codes:
        if code == 0:
            chunks.append(silence_segment(silence_s, sample_rate))
        elif code in (2, 7):
            chunks.append(sound_segment(float(code), sound_s, sample_rate, kernel))
        else:
            raise ValueError(f"Unsupported segment code {code}; use 0, 2, or 7.")
    return np.concatenate(chunks, dtype=np.float32)


DEFAULT_PLANS: List[List[int]] = [
    [0, 2, 0, 7, 0, 2, 0, 7, 0, 2],
    [0, 7, 0, 2, 0, 7, 0, 2, 0, 7],
    [0, 2, 0, 7, 0, 2, 0, 7, 0, 2],
    [0, 7, 0, 2, 0, 7, 0, 2, 0, 7],
    [0, 2, 0, 7, 0, 2, 0, 7, 0, 2],
]


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Build long WAVs from 0=silence, 2=2Hz periodic, 7=7Hz periodic blocks."
    )
    parser.add_argument(
        "--config",
        type=Path,
        default=Path(__file__).resolve().parents[2] / "config" / "stimulus_config.json",
        help="JSON with sample_rate_hz, click_duration_ms, click_amplitude, random_seed.",
    )
    parser.add_argument(
        "--outdir",
        type=Path,
        default=Path(__file__).resolve().parents[2] / "outputs" / "stimuli_sessions",
        help="Directory for session WAV files and manifest.",
    )
    parser.add_argument("--silence-s", type=float, default=240.0, help="Duration for code 0 (s).")
    parser.add_argument("--sound-s", type=float, default=120.0, help="Duration for codes 2 and 7 (s).")
    args = parser.parse_args()

    cfg = load_config(args.config)
    sr = int(cfg["sample_rate_hz"])
    click_ms = float(cfg["click_duration_ms"])
    amp = float(cfg["click_amplitude"])
    base_seed = int(cfg.get("random_seed", 42))

    args.outdir.mkdir(parents=True, exist_ok=True)

    # Same macro-sequence repeats use identical audio (A vs B template).
    seed_for_plan = [
        base_seed + 1000,  # A
        base_seed + 2000,  # B
        base_seed + 1000,
        base_seed + 2000,
        base_seed + 1000,
    ]

    manifest_path = args.outdir / "session_sequence_manifest.csv"
    with manifest_path.open("w", newline="", encoding="utf-8") as mf:
        writer = csv.DictWriter(
            mf,
            fieldnames=[
                "session_index",
                "filename",
                "path",
                "sequence_codes",
                "silence_segment_s",
                "sound_segment_s",
                "total_duration_s",
                "sample_rate_hz",
                "rng_seed",
            ],
        )
        writer.writeheader()

        for i, codes in enumerate(DEFAULT_PLANS, start=1):
            sig = concat_session(codes, args.silence_s, args.sound_s, sr, click_ms, amp, seed_for_plan[i - 1])
            name = f"session_{i:02d}_sequence.wav"
            out_path = args.outdir / name
            write_wav_mono(out_path, sig, sr)
            total_s = len(sig) / sr
            writer.writerow(
                {
                    "session_index": i,
                    "filename": name,
                    "path": str(out_path.resolve()),
                    "sequence_codes": " ".join(str(c) for c in codes),
                    "silence_segment_s": args.silence_s,
                    "sound_segment_s": args.sound_s,
                    "total_duration_s": round(total_s, 3),
                    "sample_rate_hz": sr,
                    "rng_seed": seed_for_plan[i - 1],
                }
            )
            print(f"Wrote {out_path} ({total_s/60:.2f} min)")

    print(f"Manifest: {manifest_path.resolve()}")


if __name__ == "__main__":
    main()

File: matlab/generate_session_sequence_wavs.m
function manifest = generate_session_sequence_wavs(outDir, silenceSec, soundSec, fs, clickMs, amp, baseSeed)
%GENERATE_SESSION_SEQUENCE_WAVS Long session WAVs: 0 = silence 4min, 2/7 = Hz for 2min.
%   manifest = generate_session_sequence_wavs(...)
%
% Writes five files to outputs/stimuli_sessions (by default):
%   1,3,5: [0 2 0 7 0 2 0 7 0 2]
%   2,4:   [0 7 0 2 0 7 0 2 0 7]
%
% Seeds: repeating macro-sequences share identical waveform (templates A/B).

if nargin < 1 || isempty(outDir)
    outDir = fullfile('outputs', 'stimuli_sessions');
end
if nargin < 2 || isempty(silenceSec), silenceSec = 240; end
if nargin < 3 || isempty(soundSec), soundSec = 120; end
if nargin < 4 || isempty(fs), fs = 44100; end
if nargin < 5 || isempty(clickMs), clickMs = 5; end
if nargin < 6 || isempty(amp), amp = 0.75; end
if nargin < 7 || isempty(baseSeed), baseSeed = 42; end

if ~exist(outDir, 'dir')
    mkdir(outDir);
end

planA = [0 2 0 7 0 2 0 7 0 2];
planB = [0 7 0 2 0 7 0 2 0 7];
plans = {planA, planB, planA, planB, planA};
seeds = baseSeed + [1000, 2000, 1000, 2000, 1000];

manifest = table('Size', [0 9], ...
    'VariableTypes', {'double','string','string','string','double','double','double','double','double'}, ...
    'VariableNames', {'session_index','filename','path','sequence_codes','silence_segment_s','sound_segment_s','total_duration_s','sample_rate_hz','rng_seed'});

for k = 1:numel(plans)
    rng(seeds(k));
    codes = plans{k};
    clickSamples = max(1, round(clickMs * 1e-3 * fs));
    clickKernel = amp * (2 * rand(clickSamples, 1) - 1) .* hann(clickSamples);

    y = [];
    for j = 1:numel(codes)
        code = codes(j);
        if code == 0
            ns = round(silenceSec * fs);
            y = [y; zeros(ns, 1)]; %#ok<AGROW>
        elseif code == 2 || code == 7
            ns = round(soundSec * fs);
            y = [y; make_periodic(ns, fs, double(code), clickKernel)]; %#ok<AGROW>
        else
            error('Unsupported segment code %d; use 0, 2, or 7.', code);
        end
    end

    fname = sprintf('session_%02d_sequence.wav', k);
    fpath = fullfile(outDir, fname);
    audiowrite(fpath, clamp_audio(y), fs);

    totalS = numel(y) / fs;
    seqLabel = regexprep(sprintf('%d ', codes), '\s+$', '');
    manifest = [manifest; {k, string(fname), string(fpath), string(seqLabel), ...
        silenceSec, soundSec, totalS, fs, seeds(k)}]; %#ok<AGROW>
    fprintf('Wrote %s (%.2f min)\n', fpath, totalS / 60);
end

manifestPath = fullfile(outDir, 'session_sequence_manifest.csv');
writetable(manifest, manifestPath);
fprintf('Wrote manifest: %s\n', manifestPath);
end

function y = make_periodic(nSamples, fs, hz, clickKernel)
y = zeros(nSamples, 1);
step = max(1, round(fs / hz));
for idx = 1:step:nSamples
    j = min(nSamples, idx + numel(clickKernel) - 1);
    y(idx:j) = y(idx:j) + clickKernel(1:(j - idx + 1));
end
end

function y = clamp_audio(y)
y = max(min(y, 1.0), -1.0);
end
